datasync/downloader/queue.go

```diff
@@ -873,10 +874,28 @@ func (q *queue) DeliverBodies(id string, txLists [][]*types.Transaction) (int, e
 	defer q.lock.Unlock()
 
 	validate := func(index int, header *types.Header) error {
-		if types.DeriveSha(types.Transactions(txLists[index]), header.Number) != header.TxHash {
+		txs := types.Transactions(txLists[index])
+		if len(txs) == 0 {
+			// Header-only path: avoid gov-dependent deriveSha; compare against known empty roots
+			if header.TxHash == types.EmptyTxRootOriginal || header.TxHash == types.EmptyTxRootSimple || header.TxHash == types.EmptyTxRootConcat {
+				return nil
+			}
```

Comment:
Isn't this sufficient? How often does `len(txs) = 0` and `header.EmptyBody() = false` case happen?

```
if len(txs) == 0 && header.EmptyBody() {
	return nil
}

candidates := ...
```